### PR TITLE
fix(CellBudgetFile): detect compact fmt by negative nlay

### LINKED ISSUE #1939: bug: Reading flows from binary budget file for steady stress period of zero length
I have a MF6 model consisting of two stress periods. First stress period is steady with length 0. Second stress period is transient. I can read the heads from the binary heads file with Flopy. But when I read the flow from the binary budget file, I can not read the budget for the steady stress period. The times function also doesn't return the first time. But I can read the budget data by specifying the step and period, so it is all stored. It all works fine when the steady stress period has length rather than 0. But it is nice to specify a transient stress period that starts at 0 (i.e., the first stress period has zero length).

```python
import numpy as np
import flopy as fp 

modelname = 'bugtest'
modelws = './bugtest'
sim = fp.mf6.MFSimulation(sim_name='bugtest', 
                          version='mf6', 
                          exe_name='../bin/mf6', 
                          sim_ws=modelws, 
                         )

tdis = fp.mf6.ModflowTdis(simulation=sim, 
                          nper=2, 
                          perioddata=[(0, 1, 1), (10, 10, 1)], 
                         )

ims = fp.mf6.ModflowIms(simulation=sim, 
                        complexity='SIMPLE',
                       )

gwf = fp.mf6.ModflowGwf(simulation=sim, 
                        modelname=modelname, 
                        save_flows=True,
                       )

dis = fp.mf6.ModflowGwfdis(model=gwf, 
                           nlay=1, nrow=1, ncol=10, delr=1, delc=10, top=10, botm=0
                          ) 

npf = fp.mf6.ModflowGwfnpf(model=gwf, 
                           icelltype=[0], 
                           k=10,
                          )

ic = fp.mf6.ModflowGwfic(model=gwf, 
                         strt=0,
                        )

wel = fp.mf6.ModflowGwfwel(model=gwf, 
                           stress_period_data={0: 0, 1: [[(0, 0, 0), -1]]},
                          )

sto = fp.mf6.ModflowGwfsto(model=gwf, 
                           ss=1e-4,
                           steady_state={0: True},
                           transient={1: True},
                          )

chd = fp.mf6.ModflowGwfchd(model=gwf, 
                           stress_period_data={0: [[(0, 0, 9), 0]]},
                          )

oc = fp.mf6.ModflowGwfoc(model=gwf, 
                         budget_filerecord=f"{modelname}.cbc", 
                         head_filerecord=f"{modelname}.hds", 
                         saverecord=[("HEAD", "ALL"), ("BUDGET", "ALL")],
                        )
sim.write_simulation(silent=True)
success, _ = sim.run_simulation(silent=True) 
print('success: ', success)

hds = gwf.output.head()
cbb = gwf.output.budget()
```

Then the output is:
```python
print(hds.get_times()) # gives correct times
print(hds.get_kstpkper()) # gives correct steps and periods
print(cbb.get_times()) # skips the first time
print(cbb.get_kstpkper()) # gives correct steps and periods
print(cbb.get_data(totim=0)) # gives empty list, which is wrong
print(cbb.get_data(kstpkper=(0, 0))) # returns correct values for step=0, period=0
```

### PR DESCRIPTION
**Pull request by wpbonelli**
* close #1939
  * previously, omitted totim=0 if compact
* detect compact format by negative nlay
* add mf6 test case courtesy of @mbakker7 
* test mf2005 with compact/non-compact

### CODE REVIEW

**Review by langevin-usgs — APPROVED** [review]

**Comment by wpbonelli** [comment]
Updated `CellBudgetFile` to distinguish explicitly between compact and non-compact (old-style) files by negative `nlay` values, as that seems to be the deciding factor [here](https://github.com/modflowpy/flopy/blob/637632cfa3ddbd2629dfc257759d36b851ee0063/flopy/utils/binaryfile.py#L1346). It is possible I a misinterpreting.

My understanding is that the `_totim_from_kstpkper()` method is intended for old-style non-compact files, which may or may not have time data. Previously this method was called if `totim==0` &mdash; it seems time 0 was meant to implicitly signal that a file is non-compact. This caused `_totim_from_kstpkper` to be used for compact files containing legitimate records with `totim=0`, which produced incorrect `totim` values.

**Review by langevin-usgs — APPROVED** [review]
Hey @wpbonelli, the negative nlay check looks to be more definitive (though I did not verify in the mf2005 source code itself).  The one thing that still bugs me in all this is that `_totim_from_kstpkper` should give a correct totim, even for this particular case where the first stress period has perlen set to zero -- at least that is what it appears based a quick look at that routine.

**Comment by wpbonelli** [comment]
@langevin-usgs I can come back and add some tests for `_totim_from_kstpkper()` in a separate PR, if that sounds OK.

**Comment by mbakker7** [comment]
For what it is worth, the current PR doesn't fix the problem yet of t=0 not being read from the budgetfile in MF6. Was it supposed be?

**Comment by wpbonelli** [comment]
@mbakker7 maybe I am missing something, I thought it did fix the issue. With the provided test model, `cbb.get_data(totim=0)` now returns a 4-element list, one item for each of `STO-SS`, `FLOW-JA-FACE`, `WEL`, and `CHD`, identical to what is returned by `cbb.get_data(kstpkper=(0, 0))`

```
[array([[[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]]),
 array([[[ 0.,  0.,  0., -0.,  0.,  0., -0.,  0.,  0., -0.,  0.,  0.,
         -0.,  0.,  0., -0.,  0.,  0., -0.,  0.,  0., -0.,  0.,  0.,
         -0.,  0.,  0., -0.]]]),
 rec.array([],
          dtype=[('node', '<i4'), ('node2', '<i4'), ('q', '<f8')]),
rec.array([(10, 1, 0.)],
          dtype=[('node', '<i4'), ('node2', '<i4'), ('q', '<f8')])]
```

The test failures here should resolve after https://github.com/MODFLOW-USGS/modflow6/pull/1370 is merged into mf6

**Comment by mbakker7** [comment]
I don't quite understand how the test is passing on your machine. It doesn't really on my machine. Are you sure it is passing?
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 test_read_mf6_2sp(sim)

Cell In[3], line 15, in test_read_mf6_2sp(mf6_gwf_2sp_st_tr)
     13 exp_times = [float(t) for t in range(11)]
     14 assert hds.get_times() == exp_times
---> 15 assert cbb.get_times() == exp_times
     17 # check stress periods and time steps
     18 exp_kstpkper = [(0, 0)] + [(i, 1) for i in range(10)]

AssertionError: 
```

**Comment by wpbonelli** [comment]
Just checked with mf6.4.1, 6.4.2 and latest nightly build, and all pass.. Could be a python package version issue, would you mind running `pip/conda list` and sharing output?

**Comment by mbakker7** [comment]
You really think it could be a python package version issue? That would be surpising. You really want `pip list` output? It is a long list.

**Comment by wpbonelli** [comment]
I imagine numpy is the only plausible candidate, maybe just numpy version?

File: flopy/utils/binaryfile.py [hunk 1]
```diff
@@ -1343,7 +1346,8 @@ def _get_header(self):
         """
         header1 = binaryread(self.file, self.header1_dtype, (1,))
         nlay = header1["nlay"]
-        if nlay < 0:
+        self.compact = bool(nlay < 0)
```
Comment (wpbonelli) [inline flopy/utils/binaryfile.py:1349]: for the test to fail this must be false, which it seems could happen if 

- compact cbc files don't always have negative nlay (I thought they do, but could very well be wrong)
- something is up with the type/value of `nlay` &mdash; in my debugging with numpy >= 1.24 it is a 1D, 1-element ndarray, that was the motivation to wrap with `bool()`

Comment (mbakker7) [reply flopy/utils/binaryfile.py:1349]: I was running numpy 1.23.5. I upgraded to numpy 1.26, but that did not solve the problem. It created a bunch of new error messages of the form shown below. So I am sorry to report it still doesn't work for me, and flopy needs some work for numpy 1.26 (for now adding a requirement that numpy < 1.26 until that is solved may be a good idea). 

```
/Users/markbakker/anaconda3/lib/python3.10/site-packages/flopy/utils/binaryfile.py:1692: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  ipos = int(self.iposarray[idx])
```

Comment (langevin-usgs) [reply flopy/utils/binaryfile.py:1349]: Hey @mbakker7, I checked out @wpbonelli's fix and it works for me.  I debugged through it and saw it return the correct results.  I think what is happening is that you may have added his change to an out-of-date version of flopy:develop on your system.  The reason I say that is based on the error message that you posted above.  In that error message, it references `ipos = int(self.iposarray[idx])`, which was fixed in #1868 to support new numpy requirements.  Is that possible?

image

Comment (mbakker7) [reply flopy/utils/binaryfile.py:1349]: You are right. I thought I was running the local version I have under git control on my machine, but I was running the master flopy version that I had installed. I should have realized that earlier. Sorry about wasting your time and thanks for the fix!

Comment (wpbonelli) [reply flopy/utils/binaryfile.py:1349]: @mbakker7 no problem! good news that there is no deeper issue here.

**Pull request merged**